cmd/run/logger/logger.go

```diff
@@ -0,0 +1,101 @@
+package logger
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/google/uuid"
+)
+
+var bufferSize = 1024
+var maxLineSize = 1024
+
+type Logger struct {
+	lines   []string
+	readers map[uuid.UUID]*reader
+	current int
+
+	mu sync.RWMutex
+}
+
+func New() *Logger {
+	return &Logger{
+		readers: make(map[uuid.UUID]*reader),
+		lines:   make([]string, bufferSize),
+		current: 0,
+	}
+}
+
+func (l *Logger) AddReader(readerID uuid.UUID) error {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if _, ok := l.readers[readerID]; ok {
+		return fmt.Errorf("reader with id %s already exists", readerID)
+	}
+
+	l.readers[readerID] = &reader{start: l.current, numLines: 0}
+	return nil
+}
+
+func (l *Logger) FinishRead(readerID uuid.UUID) ([]string, error) {
+	l.mu.RLock()
+	defer l.mu.RUnlock()
+
+	if r, ok := l.readers[readerID]; !ok {
+		return nil, fmt.Errorf("reader with id %s does not exist", readerID)
+	} else {
+		defer delete(l.readers, readerID)
+
+		if r.numLines <= bufferSize {
+			start := r.start
+			end := l.current
+
+			fmt.Printf("r.start=%d, r.numLines=%d, start=%d end=%d\n", r.start, r.numLines, start, end)
+			if start <= end {
+				return l.lines[start : end+1], nil
+			} else {
+				return append(l.lines[start:], l.lines[:end+1]...), nil
+			}
+		} else {
+			// We've overflowed the buffer, so just return the latest 1024 lines.
+			return append(l.lines[l.current+1:], l.lines[:l.current+1]...), nil
+		}
+
+	}
+}
+
+func (l *Logger) Write(p []byte) (int, error) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	lines := strings.Split(string(p), "\n")
+
+	for i, line := range lines {
+		if i == 0 {
+			l.lines[l.current] += line
+		} else {
+			l.lines[l.current] = line
+		}
+
+		if len(l.lines[l.current]) > maxLineSize {
+			l.lines[l.current] = l.lines[l.current][:maxLineSize]
+		}
+
+		if i != len(lines)-1 {
+			l.current = (l.current + 1) % bufferSize
+		}
+
+		for readerID := range l.readers {
+			l.readers[readerID].numLines++
```

Comment:
why track `numLines` on a per-reader basis? why not just let `l.current` be a monotonic counter and always access array elements with `l.lines[l.current % N]` instead? as a nice benefit, this would also make the writes O(1) no matter what the number of readers is.